types/react/experimental.d.ts

```diff
@@ -42,10 +42,22 @@ declare const UNDEFINED_VOID_ONLY: unique symbol;
 type VoidOrUndefinedOnly = void | { [UNDEFINED_VOID_ONLY]: never };
 
 declare module "." {
-    // Need an interface to not cause ReactNode to be a self-referential type.
-    interface PromiseLikeOfReactNode extends PromiseLike<ReactNode> {}
+    /**
+     * @internal Use `Awaited<ReactNode>` instead
+     */
+    // Helper type to enable `Awaited<ReactNode>`.
+    // Must be a copy of the non-thenables of `ReactNode`.
+    type AwaitedReactNode =
```

Comment:
A bit annoying that we have to introduce this type. TypeScript should just always flatten `Promise<Promise<T>>` to `Promise<T>` and allow `type T = U | Promise<T>`. Neither `await` nor `then()` will observe the recursive Promise.

Reply:
@rbuckton had a PR that introduced that behavior at one point, iirc, but it had some [issues in the real world](https://github.com/microsoft/TypeScript/pull/37610). The `Awaited` type as it exists today is a bit of a compromise in that regard.